zed_wrapper/src/nodelet/src/zed_wrapper_nodelet.cpp

```diff
@@ -861,31 +887,60 @@ namespace zed_wrapper {
         transformImuBroadcaster.sendTransform(transformStamped);
     }
 
-    void ZEDWrapperNodelet::publishImage(cv::Mat img,
+    void ZEDWrapperNodelet::publishImage(sl::Mat img,
                                          image_transport::Publisher& pubImg,
                                          string imgFrameId, ros::Time t) {
-        pubImg.publish(
-            imageToROSmsg(img, sensor_msgs::image_encodings::BGR8, imgFrameId, t));
+        pubImg.publish(imageToROSmsg(img, imgFrameId, t));
     }
 
-    void ZEDWrapperNodelet::publishDepth(cv::Mat depth, ros::Time t) {
-        string encoding;
-        if (openniDepthMode) {
-            depth *= 1000.0f;
-            depth.convertTo(depth, CV_16UC1); // in mm, rounded
-            encoding = sensor_msgs::image_encodings::TYPE_16UC1;
-        } else {
-            encoding = sensor_msgs::image_encodings::TYPE_32FC1;
+    void ZEDWrapperNodelet::publishDepth(sl::Mat depth, ros::Time t) {
+
+        if (!openniDepthMode) {
+            pubDepth.publish(imageToROSmsg(depth, depthOptFrameId, t));
+            return;
         }
 
-        pubDepth.publish(imageToROSmsg(depth, encoding, depthOptFrameId, t));
+        // OPENNI CONVERSION
+        sensor_msgs::ImagePtr depthMessage = boost::make_shared<sensor_msgs::Image>();
+
+        depthMessage->header.stamp = t;
+        depthMessage->header.frame_id = depthOptFrameId;
+        depthMessage->height = depth.getHeight();
+        depthMessage->width = depth.getWidth();
+
+        int num = 1; // for endianness detection
+        depthMessage->is_bigendian = !(*(char*)&num == 1);
+
+        depthMessage->step = depthMessage->width * sizeof(uint16_t);
+        depthMessage->encoding = sensor_msgs::image_encodings::MONO16;
+
+        size_t size = depthMessage->step * depthMessage->height;
+        depthMessage->data.resize(size);
+
+        uint16_t* data = (uint16_t*)(&depthMessage->data[0]);
+
+#pragma parallel for
+        for (int y = 0; y < depth.getHeight(); y++) {
+#pragma parallel for
+            for (int x = 0; x < depth.getWidth(); x++) {
+                sl::float1 value;
+                depth.getValue<sl::float1>(x, y, &value);
```

Comment:
`getValue` is not very efficient and shouldn't be used in a loop. It's better to iterate over a pointer.
Is there a data reorganization? If not it could be converted to an iteration over 1 loop